q2_cutadapt/_demux.py

```diff
@@ -122,39 +122,80 @@ def _write_empty_fastq_to_mux_barcode_in_seq_fmt(seqs_dir_fmt):
-        barcodes = pd.concat([barcodes, rev_barcodes], axis=1, sort=False)
-
-        for sample_id, f_barcode, r_barcode in barcodes.itertuples():
+    else:
+        # Quite ugly but it allows to continue to use the same logic as before
+        rev_barcodes = forward_barcodes.to_series()
+    # 'sort = false' below prevents a warning about future behavior changes
+    # by selecting the future behavior explicitly
+    barcodes = pd.concat([for_barcodes, rev_barcodes], axis=1, sort=False)
+
+    barcode_pairs = set()
+    samples_w_missing_barcodes = set()
+    samples_w_dup_barcode_pairs = set()
+    samples_w_identical_f_r = set()
+
+    for sample_id, f_barcode, r_barcode in barcodes.itertuples():
+        if reverse_barcodes is None:
+            if pd.isnull(f_barcode):
+                samples_w_missing_barcodes.add(sample_id)
+            if f_barcode in barcode_pairs:
+                samples_w_dup_barcode_pairs.add(sample_id)
+            barcode_pairs.add(f_barcode)
+        else:
             if pd.isnull(f_barcode) or pd.isnull(r_barcode):
                 samples_w_missing_barcodes.add(sample_id)
             if (f_barcode, r_barcode) in barcode_pairs:
                 samples_w_dup_barcode_pairs.add(sample_id)
             barcode_pairs.add((f_barcode, r_barcode))
-
-        if samples_w_missing_barcodes:
+            if mixed_orientation:
+                barcode_pairs.add((r_barcode, f_barcode))
+                if f_barcode == r_barcode:
+                    samples_w_identical_f_r.add(sample_id)
+
+    if samples_w_missing_barcodes:
+        if reverse_barcodes is None:
+            raise ValueError('The following samples do not have barcodes : %s'
+                             % ', '.join(sorted(samples_w_missing_barcodes)))
+        else:
             raise ValueError('The following samples do not have both '
                              'forward and reverse barcodes (note: if your '
                              'reads are in single index mixed orientation, '
                              'try again with all of your barcodes in a single '
                              'metadata column): %s'
                              % ', '.join(sorted(samples_w_missing_barcodes)))
-        if samples_w_dup_barcode_pairs:
-            raise ValueError('The following samples have duplicate barcode'
-                             ' pairs: %s' %
-                             ', '.join(sorted(samples_w_dup_barcode_pairs)))
+    if samples_w_dup_barcode_pairs:
+        raise ValueError('The following samples have duplicate barcode '
+                         '(note: if your reads are in dual index mixed'
+                         'orientation, forward-reverse pairs are also '
+                         'used as reverse-forward pairs): %s'
+                         % ', '.join(sorted(samples_w_dup_barcode_pairs)))
+    if samples_w_identical_f_r:
+        print("The following samples are using identical barcode for forward "
```

Comment:
If this branch shouldn't raise an error, it would be better to add this as a warning (vs. just a print statement).

Reply:
Should I use a simple `warnings.warn()` or is there a more "QIIMEesque" way to do so ?